File: Habitica/src/main/java/com/habitrpg/android/habitica/models/inventory/QuestContent.java
package com.habitrpg.android.habitica.models.inventory;

import android.support.annotation.Nullable;

import io.realm.RealmList;
import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;

public class QuestContent extends RealmObject implements Item {

    @PrimaryKey
    String key;
    String text, notes;
    int value, owned;
    public String previous;
    public int lvl;
    public boolean canBuy;
    public String category;
    public QuestBoss boss;

    RealmList<QuestCollect> collect;

    public String getPrevious() {
        return previous;
    }

    public void setPrevious(String previous) {
        this.previous = previous;
    }

    public int getLvl() {
        return lvl;
    }

    public void setLvl(int lvl) {
        this.lvl = lvl;
    }

    public boolean isCanBuy() {
        return canBuy;
    }

    public void setCanBuy(boolean canBuy) {
        this.canBuy = canBuy;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public QuestBoss getBoss() {
        return boss;
    }

    public void setBoss(QuestBoss boss) {
        this.boss = boss;
    }

    public RealmList<QuestCollect> getCollect() {
        return collect;
    }

    public void setCollect(RealmList<QuestCollect> collect) {
        this.collect = collect;
    }

    @Override
    public String getType() {
        return "quests";
    }

    @Override
    public String getKey() {
        return key;
    }

    @Override
    public void setOwned(int size) {
        owned = size;
    }

    @Override
    public String getText() {
        return text;
    }

    @Override
    public Integer getOwned() {
        return owned;
    }

    @Override
    public Integer getValue() {
        return value;
    }

    public void setText(String text) {
        this.text = text;
    }

    public String getNotes() {
        return notes;
    }

    public void setValue(Integer value) {
        this.value = value;
    }

    public void setNotes(String notes) {
        this.notes = notes;
    }

    public void setKey(String key) {
        this.key = key;
    }

    @Nullable
    public QuestCollect getCollectWithKey(String key) {
        for (QuestCollect collect : this.collect) {
            if (collect.key.equals(key)) {
                return collect;
            }
        }
        return null;
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/models/inventory/QuestContent.kt
package com.habitrpg.android.habitica.models.inventory

import io.realm.RealmList
import io.realm.RealmObject
import io.realm.annotations.PrimaryKey

open class QuestContent : RealmObject(), Item {

    @PrimaryKey
    override var key: String = ""
        set(value) {
            field = value
            drop?.key = value
            colors?.key = value
            boss?.key = value
        }
    override var text: String = ""
    var notes: String = ""
    var completion: String = ""
    override var value: Int = 0
    var previous: String? = null
    var lvl: Int = 0
    var isCanBuy: Boolean = false
    var category: String? = null
    var boss: QuestBoss? = null
        set(boss) {
            field = boss
            if (boss != null) {
                boss.key = key
            }
        }
    var drop: QuestDrops? = null
        set(drop) {
            field = drop
            if (drop != null) {
                drop.key = key
            }
        }
    var colors: QuestColors? = null
        set(colors) {
            field = colors
            if (colors != null) {
                colors.key = key
            }
        }

    var collect: RealmList<QuestCollect>? = null
    override var event: ItemEvent? = null

    val isBossQuest: Boolean
        get() = this.boss != null

    override val type: String
        get() = "quests"

    fun getCollectWithKey(key: String?): QuestCollect? {
        for (collect in this.collect ?: emptyList<QuestCollect>()) {
            if (collect.key == key) {
                return collect
            }
        }
        return null
    }

    fun hasGifImage(): Boolean {
        val gifImageKeys = listOf("lostMasterclasser4")
        return gifImageKeys.contains(key)
    }
}
